File: tree/1991.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node {
    char ele;
    struct Node *left;
    struct Node *right;
}Node;
Node *NewNode(char ele)
{
    Node *New;
    New = (Node*)malloc(sizeof(Node));
    New->ele = ele;
    New->left = NULL;
    New->right= NULL;
    return New;
}
Node *search_Node(Node *H, char ele)
{
    if (H != NULL) {
        if (H->ele==ele) {
            return H;
        }
        else {
            Node *tmp = search_Node(H->left,ele);
            if (tmp != NULL) {
                return tmp; 
            }
            
            return search_Node(H->right,ele);
        }
    }
    return NULL;
}
void insert_Node(Node *H,char A,char B,char C)
{
    H->ele = A;
    if (B != '.')
    {
        H->left = NewNode(B);
    }
    if (C != '.')
    {
        H->right = NewNode(C);
    }
}
 
void print_pre(Node *H)
{
    if (H!=NULL)
        printf("%c", H->ele);
    if (H->left!=NULL)
    print_pre(H->left);
    if (H->right != NULL)
    print_pre(H->right);
}
void print_in(Node *H)
{
    if (H->left!=NULL)
    print_in(H->left);
    if (H != NULL)
        printf("%c", H->ele);
    if (H->right!=NULL)
    print_in(H->right);
}
void print_post(Node *H)
{
    if (H->left != NULL)
    print_post(H->left);
    if (H->right != NULL)
    print_post(H->right);
    if (H != NULL)
        printf("%c", H->ele);
}
 
 
int main()
{
    Node *H = NewNode(NULL); //여기서는 안됨!
    Node *tmp;
    int N;
    int i;
    scanf("%d", &N);
    getchar();
    for (i = 0;i < N;i++)
    {
        char A, B, C;
        scanf("%c %c %c", &A, &B, &C);
        getchar();
        tmp = search_Node(H, A);
        if (tmp != NULL)
            insert_Node(tmp, A, B, C);
        else
        insert_Node(H,A,B,C);
    }
    print_pre(H);
    printf("\n");
    print_in(H);
    printf("\n");
    print_post(H);
}

File: tree/1991.py
# -*- coding: utf-8 -*-

# 이진 트리
'''
 작업중....
'''




class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def __str__(self):
        return str(self.data)

class Tree:
    def __init__(self):
        self.root = None

    def preorder(self, node):
        print(node, end = ' ')
        if not node.left == None : self.preorder(node.left)
        if not node.right == None : self.preorder(node.right)


    def inorder(self, node):
        if not node.left == None : self.inorder(node.left)
        print(node, end = ' ')
        if not node.right == None : self.inorder(node.right)

    def postorder(self, node):
        if not node.left == None : self.postorder(node.left)
        if not node.right == None : self.postorder(node.right)
        print(node, end = ' ')
     

    def new_node(self, element):
        new = Node()
        new.data = element
        new.left = None
        new.right = None
        return new

    def search_node(self, node, element):
        if(node != None):
            if(node.data == element):
                return node
            else:
                tmp = self.search_node(node.left, element)
                if(tmp != None):
                    return tmp
                return self.search_node(node.right, element)

    def insert_node(self, node, A, B, C):
        node.data = A

        if(B != '.'): node.left = self.new_node(B)
        if(C != '.'): node.right = self.new_node(C)


        
        


if __name__ == "__main__":


    tree = Tree()
    N = int(input())
    for _ in range(N):
        node, left_node, right_node = map(Node, input().split())
        tmp = tree.search_node()
        

    
    print(       '전위 순회 : ', end='') ; tree.preorder(tree.root)
    print('\n' + '중위 순회 : ', end='') ; tree.inorder(tree.root)
    print('\n' + '후위 순회 : ', end='') ; tree.postorder(tree.root)